gdlsc.go

```diff
@@ -1,104 +1,73 @@
 package main
 
 import (
-	"encoding/json"
 	"fmt"
-	"io/ioutil"
-	"log"
-	"net/http"
 	"os"
-	"os/exec"
+	"path/filepath"
 	"strings"
-)
-
-// URL represents path a github repos license
-var URL = "https://api.github.com/repos/%s/license?access_token=" + os.Getenv("ACCESS_TOKEN")
 
-//types for message to be printed
-type message struct {
-	Dependency string
-	License    string
-	Link       string
-}
+	"github.com/ryanuber/go-license"
+)
 
-// options for response from github
 type response struct {
-	Html_URL string
-	License  license
-}
-
-type license struct {
-	Name string
+	License string
+	File    string
 }
 
-// uses github api to fetch license based on link generated
-func getLicense(dependency string, ch chan message) {
-	url := formatLink(dependency)
-	resp, err := http.Get(url)
-	if err != nil {
-		panic(err)
-	}
-
-	defer resp.Body.Close()
-	body, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		panic(err)
-	}
-
-	cleanBody := &response{}
-	if err := json.Unmarshal([]byte(body), &cleanBody); err != nil {
-		panic(err)
-	}
-
-	link := cleanBody.Html_URL
-	license := cleanBody.License.Name
-	if link == "" {
-		link = "Not Found"
-	}
-
-	ch <- message{
-		Dependency: dependency,
-		License:    license,
-		Link:       link,
+func getLicense(input chan string, output chan response) {
+	for file := range input {
+		l, err := license.NewFromFile(file)
+		splitFile := strings.Split(file, "/")
+		newFile := splitFile[1 : len(splitFile)-1]
+		file = strings.Join(newFile, "/")
+		if err != nil {
+			output <- response{File: file, License: err.Error()}
+		} else {
+			output <- response{File: file, License: l.Type}
+		}
 	}
 }
 
-// get just the owner and repo and generate github license link
-func formatLink(link string) string {
-	strippedURL := strings.Split(link, "/")
-	authorAndRepo := strings.Join(strippedURL[1:3], "/")
-	licenseURL := fmt.Sprintf(URL, authorAndRepo)
-	return licenseURL
-}
+func main() {
+	root := "vendor/"
+	i := 0
+	input := make(chan string)
+	output := make(chan response)
 
-// removes duplicate links
-func removeDuplicates(links []string) []string {
-	linksMap := make(map[string]int)
-	for _, link := range links {
-		linksMap[link] = 1
+	for a := 0; a < 5; a++ {
+		go getLicense(input, output)
 	}
-	linksSlice := make([]string, 0, len(linksMap))
-	for link := range linksMap {
-		linksSlice = append(linksSlice, link)
-	}
-	return linksSlice
-}
+	var files []string
+	filepath.Walk(root, func(path string, f os.FileInfo, err error) error {
+		if err != nil {
+			fmt.Println("No vendor folder found. Aborting..")
+			return err
+		}
+		if !f.IsDir() {
+			splitPath := strings.Split(path, "/")
+			rawPath := splitPath[len(splitPath)-1]
+			if strings.Contains(strings.ToUpper(rawPath), "LICENSE") {
+				files = append(files, path)
```

Comment:
We also need to find a list of packages that lack a license. I think it is safe to assume that if there is a LICENSE file higher in the tree it covers all files in its subtree.